templates/_blocks/index.twig

```diff
@@ -0,0 +1,65 @@
+{# @var \Illuminate\Support\Collection<\craft\elements\Entry> blocks #}
+{% set blocks = blocks is defined ? blocks.collect() : collect([]) %}
+
+{% if blocks.count() %}
+  <div class="{{ cx({
+    'pt-64': blocks.first().type.handle not in ['pageHero', 'callToAction'],
+    'pb-64': blocks.last().type.handle not in ['callToAction'],
+  }) }}">
+  {% for block in blocks %}
+    {% set nextBlock = blocks.get(loop.index0 + 1) %}
+    {% set previousBlocks = blocks.take(loop.index0) %}
+
+    {#
+    Check if there are previous blocks that have a heading level of 1. This is used
+    to determine if the current block should use an h1 or h2 as its root heading level.
+    #}
+    {% set hasPreviousBlocksWithH1 = previousBlocks
+      |map(block => block.type.handle)
+      |intersect(['pageHero', 'heading'])
+      |length
+    %}
+
+    {% set marginBottomConfig = {
+      default: 'mb-64',
+      'heading + richText': 'mb-32',
+    } %}
+
+    {#
+    Pull margins from the config.
+
+    First we check the most unique case, the current block combined with the next block.
+
+    Such as heading + richText. Those cases are typically where we deviate from the default margin.
+
+    Next we check just the current block handle, which is useful for when we have a
+    unique case for a specific block type. Since we first check for currentBlock + nextBlock,
+    those cases will win out over the more generic ones.
+
+    Lastly we fall back to the default margin.
+
+    We use the `|filter` to remove any null values. and then `|first`
+    to get the first matching config value.
+    #}
+    {% set marginBottom = [
+      nextBlock ? "#{block.type.handle} + #{nextBlock.type.handle}" : null,
+      block.type.handle,
+      'default',
+    ]
+      |map(item => marginBottomConfig[item] ?? null)
+      |filter
+      |first
+    %}
```

Comment:
Would love any thoughts on this logic that nudges the heading block and rich text fields closer together. 

Maybe it's too complex and I should just bake a heading field into the Rich Text Block 🤷 

The idea is to sequentially check a few different aspects of block handles before. 

By checking the most specific combinations first, that should let us add cases where a specific block has unique margins when followed by a specific block AND unique padding for just that block. 

Here's an example of that scenario:

```twig
{% set marginBottomConfig = {    
    default: 'mb-64',
    'heading': 'mb-40',
    'heading + richText': 'mb-32',
} %}
```

Pretending that the block builder populated these values 

In the case that `heading` is followed by `richText`, that margin will win out. 

However, in a different scenario (say `heading + cardGrid`) there is no matching `marginBottomConfig`, so the `heading` config wins. 

```twig
{% set marginBottom = [
      'heading + richText',
      'heading',
      'default',
    ]
      |map(item => marginBottomConfig[item] ?? null)
      |filter
      |first 
%}
```